exchanges/coinbasepro/coinbasepro_websocket.go

```diff
@@ -300,61 +327,67 @@ func (c *CoinbasePro) ProcessUpdate(update *WebsocketOrderbookDataHolder, timest
 
 // GenerateDefaultSubscriptions Adds default subscriptions to websocket to be handled by ManageSubscriptions()
 func (c *CoinbasePro) generateSubscriptions() (subscription.List, error) {
-	var channels = []string{
-		"heartbeats",
-		"status",
-		"ticker",
-		"ticker_batch",
-		"candles",
-		"market_trades",
-		"level2",
-	}
-	enabledPairs, err := c.GetEnabledPairs(asset.Spot)
-	if err != nil {
-		return nil, err
-	}
-	var subscriptions subscription.List
-	for i := range channels {
-		subscriptions = append(subscriptions, &subscription.Subscription{
-			Channel: channels[i],
-			Pairs:   enabledPairs,
-			Asset:   asset.Spot,
-		})
-	}
-	return subscriptions, nil
+	return c.Features.Subscriptions.ExpandTemplates(c)
 }
 
-// Subscribe sends a websocket message to receive data from the channel
-func (c *CoinbasePro) Subscribe(channelsToSubscribe subscription.List) error {
-	chanKeys := make(map[string]currency.Pairs)
-	for i := range channelsToSubscribe {
-		chanKeys[channelsToSubscribe[i].Channel] =
-			chanKeys[channelsToSubscribe[i].Channel].Add(channelsToSubscribe[i].Pairs...)
-	}
-	for s := range chanKeys {
-		err := c.sendRequest("subscribe", s, chanKeys[s])
-		if err != nil {
-			return err
+// GetSubscriptionTemplate returns a subscription channel template
+func (c *CoinbasePro) GetSubscriptionTemplate(_ *subscription.Subscription) (*template.Template, error) {
+	return template.New("master.tmpl").Funcs(template.FuncMap{"channelName": channelName}).Parse(subTplText)
+}
+
+// Subscribe sends a websocket message to receive data from a list of channels
+func (c *CoinbasePro) Subscribe(subs subscription.List) error {
+	return c.ParallelChanOp(subs, func(subs subscription.List) error { return c.manageSubs("subscribe", subs) }, 1)
+}
+
+// Unsubscribe sends a websocket message to stop receiving data from a list of channels
+func (c *CoinbasePro) Unsubscribe(subs subscription.List) error {
+	return c.ParallelChanOp(subs, func(subs subscription.List) error { return c.manageSubs("unsubscribe", subs) }, 1)
+}
+
+// manageSub subscribes or unsubscribes from a list of websocket channels
+func (c *CoinbasePro) manageSubs(op string, subs subscription.List) error {
+	var errs error
+	subs, errs = subs.ExpandTemplates(c)
+	for _, s := range subs {
+		r := &WebsocketRequest{
+			Type:       op,
+			ProductIDs: s.Pairs.Strings(),
+			Channel:    s.QualifiedChannel,
+			Timestamp:  strconv.FormatInt(time.Now().Unix(), 10),
+		}
+		var err error
+		limitType := WSUnauthRate
+		if s.Authenticated {
+			limitType = WSAuthRate
+			err = c.signWsRequest(r)
+		}
+		if err == nil {
+			err = c.InitiateRateLimit(context.Background(), limitType)
+		}
+		if err == nil {
+			if err = c.Websocket.Conn.SendJSONMessage(r); err == nil {
+				err = c.Websocket.AddSuccessfulSubscriptions(s)
+			}
 		}
-		time.Sleep(time.Millisecond * 10)
+		errs = common.AppendError(errs, err)
 	}
 	return nil
```

Comment:
No, We want to collect all of the errors, not just return on the very first one 😄 

In reality, in situations where engine is running, we expect subs to be len(1).
However the for loop protects against a manual call to Subscribe before calling ExpandTemplates.
Also just cleaner than having to len protect subs and then use `s := subs[0]`.

Reply:
I mean at the end of the function, on line 375, we should return errs instead of nil. If we haven't collected any errors, errs will just be nil anyway, and at the end it won't interrupt execution.

Reply:
😳 
Sorry, you're completely right and I missed your point entirely
Erm. Yeah. Shocking I even brought this up to look at it 🤦 

Fixed 62c48afb